### ISSUE TITLE
aliasing.unused false positive when a table alias is used while unnesting

### ISSUE THREAD

**Issue by chronitis** [sqlfluff/sqlfluff#4623]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

[redshift] sqlfluff 2.0.2 reports (and fixes) aliasing.unused where the alias is (correctly) not being used in a select statement, but is used unnesting an array in the from statement.

Example:

```sql
-- redshift
SELECT
    my_column,
    my_array_value
FROM my_schema.my_table AS t, t.super_array AS my_array_value
```
 
```
L:   4 | P:  28 | AL05 | Alias 't' is never used in SELECT statement.
                       | [aliasing.unused]
```

### Expected Behaviour

The alias is being validly used in the FROM clause of this SELECT statement, and this rule should not trigger.

### Observed Behaviour

The table alias is reported as unused, and `sqlfluff fix` removes it, making the query invalid.

### How to reproduce

Reproducer included in _What Happened_

### Dialect

Redshift

### Version

2.0.2 (python 3.10.6)

### Configuration

Default configuration (aside from dialect).

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by pprkut** [comment]
I noticed the same thing with aliases used in MySQL `JOIN` and `WHERE` clauses (and potentially others)

**Comment by franjon** [comment]
I too have an instance where I'm using an alias in a join in the FROM statement, but not in the SELECT statement and AL05 is removing it in the FROM clause which causes the join to fail. Sometimes I need to bridge tables together with an intermediary table in the from clause but don't necessarily need to select anything from that bridge table. I've had to disable AL05 for now.

**Comment by JavierMonton** [comment]
It happens the same with `LATERAL FLATTEN` in Snowflake dialect. 

A query like this, where aliases c, d, and e are not used in the Select statement but are being used in the flatten part:

```sql
SELECT
  a.test1,
  a.test2,
  b.test3,
  f.test4
FROM table1 AS a,
  LATERAL flatten(input => some_field) AS b,
  LATERAL flatten(input => b.value) AS c,
  LATERAL flatten(input => c.value) AS d,
  LATERAL flatten(input => d.value) AS e,
  LATERAL flatten(input => e.value) AS f;
```

Fails with:
```
L:   8 | P:  40 | AL05 | Alias 'c' is never used in SELECT statement.
                       | [aliasing.unused]
L:   9 | P:  40 | AL05 | Alias 'd' is never used in SELECT statement.
                       | [aliasing.unused]
L:  10 | P:  40 | AL05 | Alias 'e' is never used in SELECT statement.
                       | [aliasing.unused]
```
And using `sqlfluff fix` removes the aliases, making the query fail.

**Comment by chronitis** [comment]
The specifically redshift-unnesting part of this looks like it was handled in https://github.com/sqlfluff/sqlfluff/pull/4775, but that probably leaves the related issues in other dialects.

**Comment by pdecat** [comment]
I believe I'm facing a related issue with `postgresql` dialect with the following query:

```sql
SELECT a
FROM json_array_elements('[{"id": "a1"}]') AS a,
    json_array_elements('[{"id": "A1"}, {"id": "A2"}]') AS b
WHERE lower(a ->> 'id') = lower(b ->> 'id');
```

```
> SELECT a
FROM json_array_elements('[{"id": "a1"}]') AS a,
    json_array_elements('[{"id": "A1"}, {"id": "A2"}]') AS b
WHERE lower(a ->> 'id') = lower(b ->> 'id');
+-------------+
| a           |
+-------------+
| {"id":"a1"} |
+-------------+
```

Running lint:
```
# sqlfluff lint -v --dialect postgres test.sql
==== sqlfluff ====
sqlfluff:                2.0.7 python:                 3.10.9
implementation:        cpython verbosity:                   1
dialect:              postgres templater:               jinja
rules:                                all
==== readout ====

== [test.sql] FAIL
L:   1 | P:   8 | RF02 | Unqualified reference 'a' found in select with more than
                       | one referenced table/view.
                       | [references.qualification]
L:   2 | P:  47 | AL05 | Alias 'a' is never used in SELECT statement.
                       | [aliasing.unused]
L:   3 | P:  60 | AL05 | Alias 'b' is never used in SELECT statement.
                       | [aliasing.unused]
L:   4 | P:  13 | RF02 | Unqualified reference 'a' found in select with more than
                       | one referenced table/view.
                       | [references.qualification]
L:   4 | P:  33 | RF02 | Unqualified reference 'b' found in select with more than
                       | one referenced table/view.
                       | [references.qualification]
==== summary ====
violations:        5 status:         FAIL
All Finished 📜 🎉!
```

Running fix removes the alias:
```
# sqlfluff fix -v --dialect postgres test.sql
==== sqlfluff ====
sqlfluff:                2.0.7 python:                 3.10.9
implementation:        cpython verbosity:                   1
dialect:              postgres templater:               jinja
rules:                                all
==== finding fixable violations ====
== [test.sql] FAIL
L:   2 | P:  47 | AL05 | Alias 'a' is never used in SELECT statement.
                       | [aliasing.unused]
L:   3 | P:  60 | AL05 | Alias 'b' is never used in SELECT statement.
                       | [aliasing.unused]
==== fixing violations ====
2 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n] ...
Attempting fixes...
Persisting Changes...
== [test.sql] FIXED
Done. Please check your files to confirm.
All Finished 📜 🎉!
  [3 unfixable linting violations found]
```
Resulting `test.sql` file content: 
```sql
SELECT a
FROM json_array_elements('[{"id": "a1"}]'),
    json_array_elements('[{"id": "A1"}, {"id": "A2"}]')
WHERE lower(a ->> 'id') = lower(b ->> 'id');
```
Which is now incorrect:
```
> SELECT a
FROM json_array_elements('[{"id": "a1"}]'),
    json_array_elements('[{"id": "A1"}, {"id": "A2"}]')
WHERE lower(a ->> 'id') = lower(b ->> 'id');

Error: table name "json_array_elements" specified more than once (SQLSTATE 42712)
```

As a work-around, I'm ignoring those rules for now:

```
-- Work-around issue where aliases are seen as unused and removed,
-- sqlfluff:warnings:AL05,RF02
```

**Comment by alanmcruickshank** [comment]
Updating this issue as at October 2023.

Confirmed that the query from the OP in redshift currently works fine ✅:
```sql
-- redshift
SELECT
    my_column,
    my_array_value
FROM my_schema.my_table AS t, t.super_array AS my_array_value
```

The two others do still fail:

```sql
-- snowflake
SELECT
  a.test1,
  a.test2,
  b.test3,
  f.test4
FROM table1 AS a,
  LATERAL flatten(input => some_field) AS b,
  LATERAL flatten(input => b.value) AS c,
  LATERAL flatten(input => c.value) AS d,
  LATERAL flatten(input => d.value) AS e,
  LATERAL flatten(input => e.value) AS f;
```

```
== [test1.sql] FAIL
L:   8 | P:  40 | AL05 | Alias 'c' is never used in SELECT statement.
                       | [aliasing.unused]
L:   9 | P:  40 | AL05 | Alias 'd' is never used in SELECT statement.
                       | [aliasing.unused]
L:  10 | P:  40 | AL05 | Alias 'e' is never used in SELECT statement.
                       | [aliasing.unused]
```

```sql
-- postgres
SELECT a
FROM json_array_elements('[{"id": "a1"}]') AS a,
    json_array_elements('[{"id": "A1"}, {"id": "A2"}]') AS b
WHERE lower(a ->> 'id') = lower(b ->> 'id');
```

```
== [test1.sql] FAIL
L:   2 | P:  47 | AL05 | Alias 'a' is never used in SELECT statement.
                       | [aliasing.unused]
L:   3 | P:  60 | AL05 | Alias 'b' is never used in SELECT statement.
                       | [aliasing.unused]
```

This first (`snowflake`) example should be fairly solvable, but the latter might be harder. In the `postgres` example, `json_array_elements` returns a column, not a table, and so when we see `FROM json_array_elements() as a`, SQLFluff assumes that `a` is a table name, not a column name. That's why why we refer to the column `a` in the `SELECT` clause - it gets confused, because it never sees a reference to a *table* called `a`. The challenge here is that SQLFluff isn't currently smart enough to tell between functions which return tables and functions which return columns. I'm not sure how best to resolve that example 🤔 .

**Comment by dbrakman** [comment]
I confirm this also affects the BigQuery dialect: running `sqlfluff fix -f --dialect=bigquery --rules=AL05` on a file containing this SQL:
```sql
with
sequences as (
  select [0, 1, 1, 2] as some_numbers
  union all
  select [2, 4, 8]
)
select
  num
from sequences as s
cross join unnest(s.some_numbers) as num
```
produces the invalid SQL below:
```sql
with
sequences as (
  select [0, 1, 1, 2] as some_numbers
  union all
  select [2, 4, 8]
)
select
  num
from sequences
cross join unnest(s.some_numbers) as num
```

**Comment by jpmckinney** [comment]
Here's a similar, even smaller test case.

* dialect = postgres
* sqlfluff, version 3.2.2

```sql
SELECT t FROM jsonb_array_elements_text('["orange","banana","watermelon"]') AS t;
```

```shell-session
$ sqlfluff fix t.sql
==== finding fixable violations ====
== [t.sql] FAIL                                                                                                                                                                                                 
L:   1 | P:  80 | AL05 | Alias 't' is never used in SELECT statement.
                       | [aliasing.unused]
== [t.sql] FIXED
1 fixable linting violations found
```

This removes the `AS t;`, which causes the query to fail:

```none
psql:t.sql:1: ERROR:  column "t" does not exist
LINE 1: SELECT t FROM jsonb_array_elements_text('["orange","banana",...
```
